src/name/ip_address.rs

```diff
@@ -14,6 +14,236 @@
 
 use crate::Error;
 
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+
+const VALID_IP_BY_CONSTRUCTION: &str = "IP address is a valid string by construction";
+
+/// Either a Ipv4 or Ipv6 address, plus its owned string representation
+#[cfg(feature = "alloc")]
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub enum IpAddress {
+    /// An ipv4 address and its owned string representation
+    IpV4Address(String, [u8; 4]),
```

Comment:
Should we mimic the `std` names here? Why not use `std::net::Ipv4Addr` in place of `[u8; 4]` (which as the same representation internally IIRC)?

Also, it might be nice to support an owned `IpAddress` with a `&'static str` instead of `String` only. Maybe use `Cow<'static, str>` instead?

Reply:
I am fine mimicking `std` names. I didn't use `std::net::Ipv4Addr` because `std` support is an optional feature.

I wanted to be strict about the no std feature, there's no other reason why I didn't use this types. If it's fine to use `std` types even when the `std` feature is disabled, we can use that instead.

Reply:
Addressed in https://github.com/rustls/webpki/pull/21.

Reply:
IIRC `IpAddr` is on a path to `core` but probably can't wait for that.